browser/net/brave_request_handler.cc

```diff
@@ -0,0 +1,298 @@
+  return ContainsKey(callbacks_, request_identifier);
+  return true;
+}
+
+int BraveRequestHandler::OnBeforeURLRequest(
+    std::shared_ptr<brave::BraveRequestInfo> ctx,
+    net::CompletionOnceCallback callback,
+    GURL* new_url) {
+  if (before_url_request_callbacks_.empty()) {
+    return net::OK;
+  }
+  ctx->new_url = new_url;
+  ctx->event_type = brave::kOnBeforeRequest;
+  callbacks_[ctx->request_identifier] = std::move(callback);
+  RunNextCallback(ctx);
+  return net::ERR_IO_PENDING;
+}
+
+int BraveRequestHandler::OnBeforeStartTransaction(
+    std::shared_ptr<brave::BraveRequestInfo> ctx,
+    net::CompletionOnceCallback callback,
+    net::HttpRequestHeaders* headers) {
+  if (before_start_transaction_callbacks_.empty()) {
+    return net::OK;
+  }
+  ctx->event_type = brave::kOnBeforeStartTransaction;
+  ctx->headers = headers;
+  ctx->referral_headers_list = referral_headers_list_.get();
+  callbacks_[ctx->request_identifier] = std::move(callback);
+  RunNextCallback(ctx);
+  return net::ERR_IO_PENDING;
+}
+
+int BraveRequestHandler::OnHeadersReceived(
+    std::shared_ptr<brave::BraveRequestInfo> ctx,
+    net::CompletionOnceCallback callback,
+    const net::HttpResponseHeaders* original_response_headers,
+    scoped_refptr<net::HttpResponseHeaders>* override_response_headers,
+    GURL* allowed_unsafe_redirect_url) {
+  if (!ctx->tab_origin.is_empty()) {
+    brave::RemoveTrackableSecurityHeadersForThirdParty(
+        ctx->request_url, url::Origin::Create(ctx->tab_origin),
+        original_response_headers, override_response_headers);
+  }
+
+  if (headers_received_callbacks_.empty()) {
+    return net::OK;
+  }
+
+  callbacks_[ctx->request_identifier] = std::move(callback);
+  ctx->event_type = brave::kOnHeadersReceived;
+  ctx->original_response_headers = original_response_headers;
+  ctx->override_response_headers = override_response_headers;
+  ctx->allowed_unsafe_redirect_url = allowed_unsafe_redirect_url;
+
+  base::PostTaskWithTraits(FROM_HERE, {content::BrowserThread::IO},
+                           base::Bind(&BraveRequestHandler::RunNextCallback,
+                                      base::Unretained(this), ctx));
+  return net::ERR_IO_PENDING;
+}
+
+void BraveRequestHandler::OnURLRequestDestroyed(
+    std::shared_ptr<brave::BraveRequestInfo> ctx) {
+  if (ContainsKey(callbacks_, ctx->request_identifier)) {
+    callbacks_.erase(ctx->request_identifier);
+  }
+}
+
+void BraveRequestHandler::RunCallbackForRequestIdentifier(
+    uint64_t request_identifier,
+    int rv) {
+  std::map<uint64_t, net::CompletionOnceCallback>::iterator it =
+      callbacks_.find(request_identifier);
+  // We intentionally do the async call to maintain the proper flow
+  // of URLLoader callbacks.
+  base::PostTaskWithTraits(FROM_HERE, {content::BrowserThread::IO},
+                           base::BindOnce(std::move(it->second), rv));
+}
+
+// TODO(iefremov): Merge all callback containers into one and run only one loop
```

Comment:
Decided to refactor this as a followup to limit amount of changes and reduce the risk of breaking everything

Reply:
can you create a ticket to track it?

Reply:
https://github.com/brave/brave-browser/issues/5574